### PR TITLE
Implement new Google two step login

### PR DESCRIPTION
**Pull request by m-soltys**
_see https://support.google.com/accounts/answer/7338427_

### CODE REVIEW

**Comment by m-soltys** [comment]
I think I also need to fix logging by clicking on the account name on the toolbar as it's not working with this two step authentication.

**Comment by ath0mas** [comment]
Seen existing issue #45 for 1-click auto connect?

**Comment by m-soltys** [comment]
Haven't seen that issue. Will fix that #45 issue under this PR.

**Review by ath0mas — CHANGES_REQUESTED** [review]
Please squash last commit with interactive rebase

**Comment by m-soltys** [comment]
@ath0mas Done.

**Review by ath0mas — DISMISSED** [review]

**Comment by ath0mas** [comment]
Supposed to work seamlessly for accounts without 2-step auth?

**Comment by m-soltys** [comment]
It's not working for the accounts without 2-step auth. I think that all accounts have it enabled so no need for supporting single step log in.

**Comment by ath0mas** [comment]
I added current snapshot version to Firefox Dev Edition and tested with 1-step auth account:
- authentication is ok, and unread counts are back to normal 👍 

but one-click auto connect from toolbar or from mail snippet fails as before (I end up on https://support.google.com/accounts/answer/61416 or on Signin page)

**Comment by ath0mas** [comment]
+ what type of 2-step verification to configure? compatible with sms, invite, ..., trusted devices, ..., with app password?

sms ko
invite ko
I don't understand how to make it work with 2-step auth at the moment

**Comment by m-soltys** [comment]
I misled you probably. By the 2-step auth I mean two steps login into Google account. Not the real two factor authentication.

I don't have a much time lately but I will try to look at non working one-click login as soon as possible. For me and couple other persons it's working just fine.

**Comment by ath0mas** [comment]
Indeed I misunderstood ;)
really great to have worked on this 2-step auth already!

(existing issues for 2-step _verification_ are #24 and #25)

**Comment by ath0mas** [comment]
I'm thinking about releasing fix for auth _quickly_ - need to review PR corresponding code before though
and _later_, with more test, 1-click autoconnect and click on snippet (as not working either for me with your xpi shared on google drive in #51)

splitting commits to different PRs ?
What do you think is better?

**Comment by bacon-cheeseburger** [comment]
I am currently testing "gmail_manager_community-0.7.5-SNAPSHOT-fx+sm.xpi" and can report that 1-click autoconnect/autologin is not working. Whenever I click on an account, I get taken to a page to enter the password for that account. Also, once I do log into an account, the new email count is always 0.

My experience is that this version of gmc isn't working at all unfortunately.

**Comment by internero** [comment]
I am using the same version, gmail_manager_community-0.7.5-SNAPSHOT-fx+sm.xpi,
and have 7 accounts: 2 @gmail.com and 5 gapps.

Since I installed this version only one account have error exactly as you
mentioned: cannot login, 0 messages, 0 bytes used and when I click the new
page opens asking for a password.

The strange thing: all other 6 accounts are working very well, with
auto-login, message counts, refresh, 1-click login... all fine!

If I can make some tests, grab some logs, and help to test, give me some
directons on what I can do.

... and thanks for the fix! Its not 100%, but it IS a improvement to keep
plugin working after Google's changes.


internero



internero@gmail.com

**Comment by m-soltys** [comment]
Guys, I'm quite busy this week. Will look at all these issues at the end of this week or beginning of next week. 

I have 4 accounts and all of them are working just fine. Maybe you have some ideas or observations why it's working for some of your account and for others not. 

It would be appreciated and help me a lot if someone could create and share with me a test account that isn't working as I don't have non working account that I can debug and see where the problem is.

File: components/gmLogger.js [hunk 1]
```diff
@@ -18,7 +18,7 @@ function gmLogger() {
 
   // Get the current debug preference value (silent)
   //this._debug = this._branch.getBoolPref("debug");
-  this._debug = true;
+  this._debug = false;
```
Comment (ath0mas) [inline components/gmLogger.js]: Keep _debug previous value here or use commented block instead?

File: chrome/content/utils/accounts.js [hunk 2]
```diff
@@ -88,7 +88,7 @@ var gmanager_Accounts = new function() {
   this.openMessage = function(aEmail, aLocation, aMessageId) {
     if (this._manager.isAccount(aEmail)) {
       var account = this._manager.getAccount(aEmail);
-      var message = account.getMessage(null, aMessageId);
+      var message = account.getMessage(account.password, aMessageId);
```
Comment (ath0mas) [inline chrome/content/utils/accounts.js]: null param is same, as processed as `(aPassword || this.password)`
so, I would not change it

File: chrome/content/utils/accounts.js [hunk 3]
```diff
@@ -74,7 +74,7 @@ var gmanager_Accounts = new function() {
   this.openFolder = function(aEmail, aLocation, aFolderId) {
     if (this._manager.isAccount(aEmail)) {
       var account = this._manager.getAccount(aEmail);
-      var folder = account.getFolder(null, aFolderId);
+      var folder = account.getFolder(account.password, aFolderId);
```
Comment (ath0mas) [inline chrome/content/utils/accounts.js]: do not change param here too

File: chrome/content/alert/alert.js [hunk 4]
```diff
@@ -174,7 +174,7 @@ var gmanager_Alert = new function() {
     document.getElementById("gmanager-alert-details-from").value = gmanager_Utils.toUnicode(snippet.from);
     document.getElementById("gmanager-alert-details-date").value = gmanager_Utils.toUnicode(snippet.time);
     document.getElementById("gmanager-alert-details-subject").value = gmanager_Utils.toUnicode(snippet.subject);
-    document.getElementById("gmanager-alert-description").firstChild.value = gmanager_Utils.toUnicode(snippet.msg);
+    document.getElementById("gmanager-alert-description").value = gmanager_Utils.toUnicode(snippet.msg);
```
Comment (ath0mas) [inline chrome/content/alert/alert.js:177]: 👍 but same change should be done on line 164 too

File: components/gmConnection.js [hunk 5]
```diff
@@ -48,6 +48,9 @@ gmConnection.prototype = {
   },
 
   send: function(aUrl, aData) {
+    // Clear data
+    this._data = "";
```
Comment (ath0mas) [inline components/gmConnection.js:52]: why need to clear data?

File: components/gmServiceGmail.js [hunk 6]
```diff
@@ -120,14 +120,21 @@ gmServiceGmail.prototype = {
...
-    var serviceURI = {
-      "url": this._loginURL,
-      "data": this._getFirstStepPostData(connection.data, (aPassword || this._password), aContinueData)
-    };
+    var data = connection.data;
+    var postUrl = this._extractActionUrlFromFormData(data);
+    connection.send(postUrl, this._getFirstStepPostData(data), this);
+
+    data = connection.data;
+    postUrl = this._extractActionUrlFromFormData(data);
```
Comment (ath0mas) [inline components/gmServiceGmail.js:128]: duplicated lines 127-128 / 123-124?

File: components/gmServiceGmail.js [hunk 7]
```diff
@@ -120,14 +120,21 @@ gmServiceGmail.prototype = {
     var connection = Components.classes["@gmail-manager-community.github.com/gmanager/connection;1"].createInstance(Components.interfaces.gmIConnection);
     connection.send(this._loginURL, null);
 
-    var serviceURI = {
-      "url": this._loginURL,
-      "data": this._getFirstStepPostData(connection.data, (aPassword || this._password), aContinueData)
-    };
+    var data = connection.data;
```
Comment (ath0mas) [inline components/gmServiceGmail.js]: really need to extract this `data` var or should be used directly with `connection.data` like on line 135?

File: components/gmServiceGmail.js [hunk 8]
```diff
@@ -233,10 +240,17 @@ gmServiceGmail.prototype = {
...
 
-    aCookies.forEach(function(cookie, index, array) {
+    aCookies.forEach(function(cookie) {
       this._log("cookie name = " + cookie.name);
       this._log("cookie value = " + cookie.value);
-      this._cookieManager.add(cookie.host, cookie.path, cookie.name, cookie.value, cookie.isSecure, cookie.isHttpOnly, cookie.isSession, cookie.expires);
+
+      // If cookie.expires equals to 0 then the cookie should be set as a session cookie. It doesn't happen so need to set expiration time manually.
+      // Expiration time is in seconds from epoch, now will be 60 seconds from now
+      var cookieExpirationTime = !cookie.expires
```
Comment (ath0mas) [inline components/gmServiceGmail.js]: I would not define new variable here but only change `cookie.expires` value as you do, to use it inside _cookieManager.add as before

**Review by ath0mas — CHANGES_REQUESTED** [review]

File: components/gmLogger.js (hunk 1 above)
Comment (m-soltys) [reply components/gmLogger.js]: Will use commented block instead, there is no need to have log turned on in released instances.

File: components/gmConnection.js (hunk 5 above)
Comment (m-soltys) [reply components/gmConnection.js:52]: As you can see in lines 69-71 there is appending the stream data to the _data variable. If you're using the send function on the same instance of gmConnection twice then data from the second request is appended to the data from the first request which is bad.

File: components/gmServiceGmail.js (hunk 6 above)
Comment (m-soltys) [reply components/gmServiceGmail.js:128]: Did you miss the `connection.send` between duplication? The second part of code is working on the different response.  

The flow:
1. Get the login page from server - line 121
2. Parse the response and get required data - lines 123-124
3. Post request to the server with account name (1st step login) - line 125
4. Parse the response and get required data - 127-128

File: chrome/content/utils/accounts.js (hunk 2 above)
Comment (m-soltys) [reply chrome/content/utils/accounts.js]: Don't get it. Could you explain? 

I need to pass the password to the `_getServiceURI` funciton as this._password was null for me.

File: chrome/content/utils/accounts.js (hunk 3 above)
Comment (m-soltys) [reply chrome/content/utils/accounts.js]: As explained.

File: chrome/content/alert/alert.js (hunk 4 above)
Comment (m-soltys) [reply chrome/content/alert/alert.js:177]: Done.

File: components/gmServiceGmail.js (hunk 8 above)
Comment (m-soltys) [reply components/gmServiceGmail.js]: My bad, missed that I'm changing `cookie.expires`. Will fix to use only the new variable. 

Shouldn't change function's argument.

File: chrome/content/utils/accounts.js (hunk 2 above)
Comment (ath0mas) [reply chrome/content/utils/accounts.js]: Here it is `getMessage` from gmAccount not gmServiceGmail

File: chrome/content/utils/accounts.js (hunk 3 above)
Comment (ath0mas) [reply chrome/content/utils/accounts.js]: same 😋 
and see lines 102-104 with similar structure `account.getCompose(null, aHref)` (all 3 called with `null`)

File: components/gmServiceGmail.js (hunk 6 above)
Comment (ath0mas) [reply components/gmServiceGmail.js:128]: fine, I did not see getter for `_data` behind `connection.data`

File: components/gmConnection.js (hunk 5 above)
Comment (ath0mas) [reply components/gmConnection.js:52]: fine

File: components/gmServiceGmail.js (hunk 8 above)
Comment (ath0mas) [reply components/gmServiceGmail.js]: ok ; maybe write it in _positive_ and short way: `var cookieExpirationTime = cookie.expires || Math.floor(new Date() / 1000) + 60;`

File: components/gmLogger.js (hunk 1 above)
Comment (ath0mas) [inline components/gmLogger.js]: can remove line 21?

File: components/gmLogger.js [hunk 9]
```diff
@@ -17,8 +17,8 @@ function gmLogger() {
   this._branch.addObserver("", this, false);
 
   // Get the current debug preference value (silent)
-  //this._debug = this._branch.getBoolPref("debug");
-  this._debug = true;
+  this._debug = this._branch.getBoolPref("debug");
+  this._debug = false;
```
Comment (ath0mas) [inline components/gmLogger.js]: can remove line 21?

**Review by ath0mas — CHANGES_REQUESTED** [review]

File: chrome/content/utils/accounts.js (hunk 2 above)
Comment (m-soltys) [reply chrome/content/utils/accounts.js]: Ahhh I see it now, you're right.

File: components/gmServiceGmail.js (hunk 8 above)
Comment (m-soltys) [reply components/gmServiceGmail.js]: I will never get used to JS :)

File: components/gmLogger.js (hunk 9 above)
Comment (m-soltys) [reply components/gmLogger.js]: Done.

**Review by ath0mas — APPROVED** [review]

**Pull request merged**